{"cells":[{"cell_type":"markdown","metadata":{"id":"67eZMV04GDd9"},"source":["## ORF finding\n","\n","## Deliverables:\n"," - Lab5 notebook with complete inspection cells ( Introduction and Results )\n"," - findORFs.py - 50 total points\n"," - imports\n","     - sequenceAnalysis (for FastaReader at least )\n","     - your OrfFinder class ( placed in sequenceAnalysis )\n","     \n","## required output files:\n"," - tass2ORFdata-ATG-100.txt\n","\n","## classes required (minimally):\n"," - OrfFinder\n"," - FastAreader\n","\n","## possible extra credit options - 10 additional points\n"," - minOrf= [integer]  This allows exclusion of genes with lengths < minOrf  (Note this length includes both start and stop codons)\n"," - biggestOrfOnly True | False  This allows reporting of all genes or only the largest within an ORF\n"," - startCodons= ['ATG' | 'TTG' | 'GTG']  This parameter produces a list of start codons used in determining genes ( you might convert this to a more useful container like a set.)\n"," - stopCodons = ['TAA' | 'TGA' | 'TAG']  This parameter specifies which codons are to be treated as valid stop codons\n","\n","## Due: Monday February 19, 2024 11:59pm\n","\n","# Overview\n","This lab will give you practice with a real algorithm and with formatted output. We will be finding genes in a genome.\n","\n","An open reading frame (ORF) is a region of DNA that is free from stop codons (TAG, TAA, TGA). We can then define a putative gene in that ORF by the presence of a start codon at the 5' end of the ORF and a stop codon at the 3' end. Start codons in some bacteria, archaea and viruses can be ATG, GTG or CTG, and in some cases TTG.  \n","\n","Since genes are made up of codons of three symbols, any particular sequence of DNA can be interpreted in each of 3 overlapping coding frames, and since cellular genomic DNA is double stranded, we can have genes located on both strands, yielding 6 possible coding frames for the many genes in a genome.  Genes can also overlap each other.\n","\n","To limit the number of possible genes that we might find, two basic strategies are sometimes used:\n","\n"," - look for genes that are longer than some predefined number (100 bases for example) and\n"," - only consider the largest coding region within an ORF, rather than considering all substrings in the ORF that begin with intermediate start codons.\n","\n","A well known ORF finder is available at NCBI - ORF Finder https://www.ncbi.nlm.nih.gov/orffinder/\n","\n","In this assignment, you will write an equivalent tool that provides the same function, though without the graphical display.\n","\n","Use your sequenceAnalysis module in designing a solution.\n","\n","Your design is essential for success.  Spend time considering how best to solve this problem.  Write out pseudocode that describes the core algorithm.  This pseudocode is required to be delivered in comments at the top of your code submission.  Please start early on this assignment, and consider the algorithm carefully.  \n"," - Start and stop codons must be in the same frame,\n"," - many possible Met codons (or other possible starts) may exist in an ORF, and\n"," - genes can have significant overlap.\n"," - In some cases, 2 or more ORFS can exist on opposite strands in the same coordinate space.\n","\n","A partial solution exists using regular expressions.  This solution does not handle any of the extra-credit possibilities, and does not handle the required boundary conditions involving gene fragments. Feel free to look, though I did not find this approach helpful. It also uses regular expressions, which are both difficult to read and disallowed for this assignment.\n","\n","Your implementation should be designed to add capability to your sequenceAnalysis module. Please carefully consider where that capability belongs. You are required to make use of your sequenceAnalysis module for most of the code and for FastAreader. You might consider adding  a function to calculate reverse complement. You should also add an OrfFinder class to sequenceAnalysis. For this assignment, you are the designer of your toolbox - choose wisely. You might use this again in your final project.\n","\n","# ORF finder\n","Create a program to analyze a FASTA-formatted file containing a sequence of DNA and find the ORFs (start and stop codons). Your program is called findORFs.py and it:\n","\n"," - reads in FASTA-formatted sequence data from a file using STDIN\n"," - finds ORFs that are >= 100 nucleotides in length\n"," - writes formatted output data to a text file using STDOUT\n"," - if there are multiple FASTA-formatted sequences in the file, you need to perform Orf-finding on each of them independently.\n","\n","# Design specification\n","\n","## Input file\n","Input data must make use of STDIN. This will allow commandLine execution using:\n","- findOrfs.py < someInput.\n","I will provide 2 test files for your use.  For debugging purposes, use lab5test.fa, which is provided in Canvas::files.  As your final test, use the tass2.fa genome. I have provided the output generated by my program for your comparison.\n","\n","## Output file\n","\n","Write your tass2 output to a file named: tass2ORFdata-ATG-100.txt using STDOUT.  This file should be generated with the following program features:\n","\n"," - The minimum gene size to report is 100 nucleotides (including stop and start)\n"," - The only start codon to consider is ATG\n"," - The stop codons in DNA form are: TAA TGA TAG\n"," - Only the largest putative gene in an ORF is reported\n"," - The input file is tass2.fa\n","\n","I have provided the output generated using a minimum gene size of 300, which you can use for testing.\n","\n","## Gene finding\n","\n","Your  program should find genes in each of the six possible coding frames.  The genes that you find will begin with a start codon and end with a stop codon, with the exception of those gene-fragments that might be located on the ends of the sequence (see boundary conditions).\n","\n","## Boundary conditions\n","\n","You may find that an ORF exists at either or both termini of the sequence you are examining.  This may mean that you will only be seeing a gene fragment located at the sequence termini.\n","\n","For example, consider the sequence:<br>\n","AAA AAA AAA TGA CCC CCC ...\n","\n","Here, we see a stop codon with a set of AAA codons located upstream.  In this case, assume that a start codon exists upstream of our sequence, and report this as a gene in Frame 1, starting at position 1, ending at position 12 and having a length of 12.  \n","\n","The same may occur on the 3' end of the sequence, for example: <br>\n","A AAA AAA ATG CCC CCC CCC CC  \n","\n","In this case, report a gene in Frame 2 starting at position 8, ending at position 21 and having a length of 14.  \n","\n","And, of course genes can exist on the opposite strand, so: <br>\n","TTA AAA AAA AAA CAT CC<br>\n","would be reported in Frame -3, starting in position 1, ending at position 15, with length 15.\n","\n","NCBI ORF finder is a good test case to see what your output should look like.  THe NCBI program does not handle this boundary case however.\n","\n","## Sort Order\n","\n","Your output should be sorted by decreasing ORF size, and in cases where multiple ORFs have the same size, sort those by the left position of the gene.  (build a key that makes this easy).\n","\n","## Minimum gene size (option)\n","\n","The minimum gene size to report is 100 nucleotides.  This count of nucleotides includes the start codon and stop codon.\n","\n","## Largest ORF (option)\n","\n","In any ORF, there may be multiple start codons present.  For the default assignment, only consider the largest gene in an ORF. For the extra-credit assignment, if this (largest ORF) option is not set, report all ORFs found (subject to the minimum size option).\n","\n","## Start codon(s) (option)\n","\n","For the default part of the program, only consider ATG as a start codon. For extra credit, use an input parameter that specifies a set of start codons.\n","\n","## Stop codon(s) (option)\n","\n","For the default part of the program, only consider TAG, TAA and TGA as stop codons. For extra credit,  use an input parameter that specifies the set of stop codons.\n","\n","## Report output\n","\n","All positions reported use the coordinate position of the top strand, and the first position in the sequence (on the left) is position 1.  This coordinate system is used for all 6 coding frames.\n","\n","Frames 1 through 3 can be determined using the start position of the gene.  Assuming the position is p, (p mod 3) + 1 will provide the coding frame.  For genes on the bottom strand, the frame is based on the right hand side of the sequence (see below).\n","\n","For each putative gene, your report must provide:\n","\n"," - coding frame. for the top strand, a gene starting in the first position in the sequence is in frame 1.  For the bottom strand, frames are numbered based on the right-hand coordinate position of the sequence.  So, if the sequence is 100 bases long, and you have a gene in positions 89..100, that gene is in frame -1. You can still use the coordinate positions, but the calculation is:  - ( (len(sequence) - last position) mod 3 + 1).  Remember that the last position in this case corresponds to the start codon of the gene on the bottom strand.  These conventions make sense if you think about the reverse complement of the original sequence. The coding frame should be reported with a + or - as needed to describe the coding frame\n"," - start position. This position should correspond to the first base of the start codon (top strand) or last base of the stop codon (bottom strand), or the terminus of the genome if this is a gene fragment.\n"," - end position. This position should correspond to the first base of the start codon (bottom strand) or last base of the stop codon (top strand), or the terminus of the genome if this is a gene fragment.\n"," - length. This includes the start and end codon\n","\n","\n","Line 1 - the header line from the original sequence (without >)\n","\n","Lines 2 .... (use '{:+d} {:>5d}..{:>5d} {:>5d}' or equivalent)"]},{"cell_type":"markdown","metadata":{"id":"f2S3czxpGDd_"},"source":["+1 57166..61908  4743 <br>\n","..."]},{"cell_type":"markdown","metadata":{"id":"RqALBONVGDeA"},"source":["# Extra Credit - 10 points possible\n"," - Allow your program to optionally use a set of start codons - ATG, GTG, TTG for example.\n"," - Allow your program to optionally use a set of stop codons - TAG, TGA, TAA for example.\n"," - Allow your program to print every putative gene in an ORF instead of only the largest.\n"," - Allow your program to have varying ORF size minimums\n","\n","# Submit your code and answers\n","For this lab, you should upload the following three files as attachments:\n","\n"," - your sequenceAnalysis.py module\n"," - findORFs.py, include design in # comments at the beginning of findORFs.py\n"," - tass2ORFdata-ATG-100.txt\n","\n","Important: to get full credit on this lab assignment, each of the code files you submit needs to:\n"," - Run properly (execute and produce the correct output)\n"," - Include an overview about what your program is designed to do with expected inputs and outputs\n"," - Include design (place at the top of your file below your name, group, and description). Include any assumptions or design decisions you made in writing your code\n"," - Adhere to the Report format specification\n"," - Include docstrings for Program, Module, Class, Methods and Functions\n"," - Contain in-line comments\n","\n"," Congratulations, you finished your fifth lab assignment!\n","\n","# Hints:\n","## One strand at a time\n","\n","One design solution considers only the top strand to generate gene candidates. We then generate the reverse complement and a second list of gene candidates.  When done in this way, remember to fix the start/end coordinates that you get from the reverse complement solution, since those will be based on the other end of the sequence.\n","\n","An alternative design exists for the reverse strand that does not require generation of the reverse complement. This solution is interesting and may be simpler to consider. This solution scans the bottom strand left to right, finding stop codons first (reverse complement), then looking for start codons (reverse complement).\n","\n","A third solution for the bottom strand would scan right to left, finding starts until a stop is encountered.\n","\n","## Start and Stop handling\n","\n","If you are going to do the extra-credit, then you will need to remember where the starts are until you find a stop.  This is straightforward if you place those start positions in a list, organized by frame. Even if you are not doing the extra-credit, placing starts on a list organized by frame will really simplify your code. This solution would require three lists for each of the three reading frames.\n","\n","As an alternative, you could scan each of the three reading frames seperately, counting by threes in each case. This seems to be the most straightforward way to do this assignment.\n","\n","When you do find a stop, you then consider the start(s) that you found along the way in this ORF.  Notice that the \"longest\" ORF is then defined by the start that is at the beginning of your start list.  Don't forget to clean up this list after you find genes using this stop since it would terminate all of those \"pending\" genes. What would happen if you didn't clean up your start list - you would be finding ORF candidates that had a stop in the middle of them, which means they are not __OPEN__ reading frames.\n","\n","## Termini and gene fragments\n","\n","This is a bit easier if you think that position 0(1 externally) will allways be a start position for a gene in every frame. This is the case where a start codon exists somewhere upstream of our sequence. It would code for a gene that would terminate in our sequence somewhere. This is the classic example of a dangling-stop. We would report this gene fragment beginning at position 0(1) and ending at the end of that in-frame stop. In frame 1, there might actually be a start codon at this position so dont count it twice. In Frames 2 or 3, there are extra bases ahead of our first full codon, so we will always have the possibility that the actual start is upstream of this dna fragment, so.. we report this as a dangling stop case starting at position 0(1).  \n","\n","Dangling starts always end at the end of sequence, even if that position is out of frame. In cases where no start or stop codon is found anywhere in the frame, the entire sequence is then considered as a gene.\n","\n","# Data and output files\n","\n","The cell below is set up for the non-extra credit version of the program with a fixed filename. Your much more usable program will use STDIN and STDOUT in command line fashion. To do this, simplify main so your fastaReader object is reading from STDIN.\n","\n","All of the input files and programs should be in the same folder so you won’t have to specify a path, if your data file is in some other folder, you can provide that like this:<br> ~/Desktop/bme160/someOtherPlace/labData.fa.\n","\n","# Comments\n","\n","To get full credit for this lab, you must turn in your design for each program you submit. Remember that any item you write or include must have proper docstrings.\n","\n","\n","Mona: Hi Emmet! I know how long it took you to do this lab so I'm so glad that you got it to work. Yay you did it! I thought it was smart to test with having 300 for your minGene in the def___init___ itself. I also think that it was interesting that you coded out the reverse, I can see what you were thinking as you were writing each line of code. One thing that I think would be more efficient is maybe simplifying your for loops in find Orfs, honestly, this is something I struggled with too, so maybe next time we can figure out something together. Great job :).\n","\n","Clare: Hi Emmet! Great job on this lab. I think your approach to the code is super interesting, especially the loop you have for frame finder. I think it could've been broken up a little into seperate cases but I think you made it work and it's quite an impressive loop. Overall, great code. The use of comments and docstrings made it very readable even tho the code itself was not super intuitive.\n","\n","Liam - Hey Emmet, I think the design of your code is super interesting. You have one method which returns all of the frames for one direction, a different method which returns the reverse, and final methods to attach the frames as attributes and then print them. I think this was a super interesting approach and very creative. One thing I also appreciate, like Clare and Mona, is the comments to explain the functions and every line of code. The framefinder class is very dense and the comments really help to explain it all. I personally think the way you have done it is more efficient than mine, as it does all of the steps simultaenously like finding frames, codons, and such. Great job Emmet!"]},{"cell_type":"markdown","metadata":{"id":"Je30scjIGDeA"},"source":["ORFFINDER CLASS"]},{"cell_type":"markdown","source":[],"metadata":{"id":"iCn_XHLiux-E"}},{"cell_type":"markdown","source":[],"metadata":{"id":"y2kyUUtguso_"}},{"cell_type":"code","execution_count":null,"metadata":{"id":"7WwekT3sGDeB"},"outputs":[],"source":["class OrfFinder():\n","    \"\"\"\n","    Class to represent a sequence of DNA and find gene candidates. This code has methods to find candidates on all 3 frames in the positive and negative strand and to print the results in an approriate format\n","    \"\"\"\n","    pairs = {\"A\":\"T\",\"C\":\"G\",\"T\":\"A\",\"G\":\"C\"}\n","\n","    def __init__ (self,seq,longGene=True,minGene=300,start=[\"ATG\"],stop=[\"TAA\"]):\n","        \"\"\"\n","        intializes a class with attributes for all extra credit options, and sequence and uses these values to find open reading frames and add them to an attribute\n","        \"\"\"\n","        self.start = start\n","        self.stop = stop\n","        self.frames = []\n","        self.seq = seq\n","        self.minGene = minGene\n","        self.longGene = longGene\n","        self.addseq(seq,longGene,minGene,start,stop)\n","\n","    def frameFinder(self,seq,strandpos):\n","        \"\"\"\n","        Returns a list of Open reading frames on all 3 possible frames from a sequence of DNA either the positive or negative strand\n","        \"\"\"\n","        start1 = self.start\n","        stop = self.stop\n","        frames = []\n","        for frame in range(0,3): #looping through 3 possible frames\n","            stopfound = False #tracks whether a stop codon has been encountered in this frame\n","            startpos=[] #list of start codons found\n","            for index in range(frame,len(seq)-2,3): #loops through 3 bases at a time using the frame# from the parent loop as an offset to search the right frame\n","                if seq[index:index+3] in start1: #checking for start codons\n","                    if index != 0: #preventing duplicate frames when we handle hanging cases later\n","                        startpos.append(index)\n","                if seq[index:index+3] in stop: #checking for stop coodns\n","                    startpos = sorted(startpos)\n","                    for start in startpos: #adding frames for all start positions found before this stop codon\n","                        if ((index+3)-start) >= self.minGene:\n","                            if strandpos == True:\n","                                frames.append([frame+1,start+1,index+3,(index+3)-start])\n","                            elif strandpos == False:\n","                                frames.append([-(frame+1),len(seq)-(index+2),len(seq)-start,(index+3)-start])\n","                        if self.longGene == True: #if longest gene is true we shoudl only append the frame found for the earliest start position before the stop coodn\n","                            break\n","                        #exit start positons loop all start positions prior to this stop handled, hanging cases adressed in block below\n","                    if stopfound == False: #block to handle hanging cases with no start position, if a stop has been found before this one it can not be a no start hang because a stop cant be read through\n","                        if startpos == []:\n","                            if (index+3)>=self.minGene:\n","                                if strandpos == True:\n","                                    frames.append([(frame+1),1,index+3,index+3])\n","                                elif strandpos == False:\n","                                    frames.append([-(frame+1),index+3,len(seq),index+3])\n","                    #reset start positions and update stop codon tracker to found\n","                    startpos = []\n","                    stopfound=True\n","                    #exit stop found conditional block, all start positons prior to this stop codon and any relevant hanging cases handled\n","                #exit loop iterating through codons in frame, all codons in frame searched\n","            if stopfound == False: #block to handle hanging cases for a frame with no stops\n","                for start in startpos:\n","                    if strandpos == True:\n","                        frames.append([frame+1,start+1,len(seq),len(seq)-start,])\n","                    elif strandpos == False:\n","                        frames.append([-(frame+1),1,len(seq)-start+1,len(seq)-start])\n","                if startpos == []:\n","                        if strandpos == True:\n","                            frames.append([(frame+1),1,len(seq),len(seq)])\n","                        elif strandpos == False:\n","                            frames.append([-(frame+1),1,len(seq),len(seq)])\n","            #exit parent loop, iterate to next frame if relevant\n","        frames = sorted(frames, key=lambda x:(-x[3]))\n","        return frames\n","\n","    def findReverse(self,str):\n","        \"\"\"\n","        Returns the complimentary seqeunce to a strand in the 5'-3' direction\n","\n","        input: ACTG\n","        Ouput: CAGT\n","        \"\"\"\n","        compliment = \"\"\n","        for x in str:\n","            for char in x:\n","                compliment += self.pairs[char]\n","        return compliment[::-1]\n","\n","    def addseq(self,seq,longGene,minGene,start,stop):\n","        \"\"\"\n","        Takes in a sequence and adds the frames on both reverse and forward strand to the frames attributes\n","        \"\"\"\n","        seqReverse = self.findReverse(seq) #find reverse compliment in 5-3 directio\n","        self.frames = self.frames + (self.frameFinder(seq,True)) #find frames on positive strand\n","        self.frames = self.frames + (self.frameFinder(seqReverse,False)) #find frames on reverse strand\n","\n","    def printer(self):\n","        \"\"\"\n","        prints all frames in appropriate format\n","        \"\"\"\n","\n","        frames = sorted(self.frames, key=lambda x: -x[3])\n","\n","        max_length = max(len(str(item[2])) for item in frames)\n","\n","        # Print the items with consistent spacing\n","        for item in frames:\n","            print(f\"{item[0]:+} {item[1]:<6}..{item[2]:<6}  {item[3]:<{max_length}}\")\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"ExqpcNjXGDeC"},"outputs":[],"source":["MAIN"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"XrtwEPwYGDeC","executionInfo":{"status":"error","timestamp":1708586606002,"user_tz":480,"elapsed":14,"user":{"displayName":"Emmet Benaryeh","userId":"00375674251069814643"}},"outputId":"2462cc12-5863-43e1-cb45-a180e6f25704","colab":{"base_uri":"https://localhost:8080/","height":383}},"outputs":[{"output_type":"error","ename":"ModuleNotFoundError","evalue":"No module named 'SequenceAnalysisCopy'","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)","\u001b[0;32m<ipython-input-1-e777c409f995>\u001b[0m in \u001b[0;36m<cell line: 10>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0;31m########################################################################\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0msys\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 10\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0mSequenceAnalysisCopy\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mFastAreader\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mOrfFinder\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     11\u001b[0m \u001b[0;32mclass\u001b[0m \u001b[0mCommandLine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m     '''\n","\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'SequenceAnalysisCopy'","","\u001b[0;31m---------------------------------------------------------------------------\u001b[0;32m\nNOTE: If your import is failing due to a missing package, you can\nmanually install dependencies using either !pip or !apt.\n\nTo view examples of installing some common dependencies, click the\n\"Open Examples\" button below.\n\u001b[0;31m---------------------------------------------------------------------------\u001b[0m\n"],"errorDetails":{"actions":[{"action":"open_url","actionText":"Open Examples","url":"/notebooks/snippets/importing_libraries.ipynb"}]}}],"source":["#!/usr/bin/env python3\n","# Name: Your full name (cruzid account )\n","# Group Members: List full names (cruzid usernames) or “None”\n","\n","\n","########################################################################\n","# CommandLine\n","########################################################################\n","import sys\n","from SequenceAnalysisCopy import FastAreader, OrfFinder\n","class CommandLine() :\n","    '''\n","    Handle the command line, usage and help requests.\n","\n","    CommandLine uses argparse, now standard in 2.7 and beyond.\n","    it implements a standard command line argument parser with various argument options,\n","    a standard usage and help.\n","\n","    attributes:\n","    all arguments received from the commandline using .add_argument will be\n","    avalable within the .args attribute of object instantiated from CommandLine.\n","    For example, if myCommandLine is an object of the class, and requiredbool was\n","    set as an option using add_argument, then myCommandLine.args.requiredbool will\n","    name that option.\n","\n","    '''\n","\n","    def __init__(self, inOpts=None) :\n","        '''\n","        Implement a parser to interpret the command line argv string using argparse.\n","        '''\n","\n","        import argparse\n","        self.parser = argparse.ArgumentParser(description = 'Program prolog - a brief description of what this thing does',\n","                                             epilog = 'Program epilog - some other stuff you feel compelled to say',\n","                                             add_help = True, #default is True\n","                                             prefix_chars = '-',\n","                                             usage = '%(prog)s [options] -option1[default] <input >output'\n","                                             )\n","        self.parser.add_argument('-lG', '--longestGene', action = 'store', nargs='?', const=True, default=False, help='longest Gene in an ORF')\n","        self.parser.add_argument('-mG', '--minGene', type=int, choices= (0,100,200,300,500,1000), default=300, action = 'store', help='minimum Gene length')\n","        self.parser.add_argument('-s', '--start', action = 'append', default = ['ATG'],nargs='?',  help='start Codon') #allows multiple list options\n","        self.parser.add_argument('-t', '--stop', action = 'append', default = ['TAG','TGA','TAA'],nargs='?', help='stop Codon') #allows multiple list options\n","        self.parser.add_argument('-v', '--version', action='version', version='%(prog)s 0.1')\n","        if inOpts is None :\n","            self.args = self.parser.parse_args()\n","        else :\n","            self.args = self.parser.parse_args(inOpts)\n","\n","########################################################################\n","# Main\n","# Here is the main program\n","#\n","#\n","########################################################################\n","\n","def main(inFile = None, options = None):\n","    '''\n","    Takes in an infile name and a series of options and returns all gene candidates that fit the options\n","    '''\n","    thisCommandLine = CommandLine(options) #initializing commandline object to handle the options input from the command line\n","    reader = FastAreader(inFile) #fasta read the infie\n","    for head,seq in reader.readFasta(): #print all gene candidates in each fasta sequence fragment in the file\n","        print(head)\n","        myOrfFinder = OrfFinder(seq,thisCommandLine.args.longestGene,thisCommandLine.args.minGene,thisCommandLine.args.start,thisCommandLine.args.stop)\n","        myOrfFinder.printer()\n","\n","if __name__ == \"__main__\":\n","    main(inFile='tass2.fa', options=['-mG=300'])\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"J3arUAbHGDeD"},"outputs":[],"source":["python3 findOrfs.py  <somefile  >someOtherFile -mG 300 -lG -s ATG -s GTG"]},{"cell_type":"markdown","metadata":{"id":"CtEMlLqzGDeD"},"source":["# findORFs output using tass2.fa  -mG=300 -lG"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"jV9q9czbGDeE"},"outputs":[],"source":["tass2 NODE_159_length_75728_cov_97.549133\n","+1 57166..61908  4743\n","-1  8192..11422  3231\n","+2 65963..69004  3042\n","-3 14589..16862  2274\n","-2  2968.. 4872  1905\n","+1 64093..65952  1860\n","-3    30.. 1694  1665\n","+1 69475..71052  1578\n","+1 48805..50223  1419\n","+1 47398..48798  1401\n","-3 29133..30500  1368\n","-1 40922..42250  1329\n","-2 19975..21270  1296\n","+3 72273..73559  1287\n","-1 24482..25639  1158\n","-1  6689.. 7804  1116\n","+1 50251..51366  1116\n","-3 27501..28601  1101\n","+2 63038..64078  1041\n","+3 62019..63038  1020\n","-2 42271..43263   993\n","+3 51864..52805   942\n","+1 45484..46371   888\n","-3 11433..12317   885\n","+2 74357..75184   828\n","+2 71576..72394   819\n","+3 46341..47138   798\n","-2 18613..19407   795\n","+1 55642..56388   747\n","-1 16940..17632   693\n","-2 13288..13974   687\n","-3 26115..26801   687\n","-1 21338..21994   657\n","-1 30998..31654   657\n","-2 12601..13251   651\n","-2  4894.. 5532   639\n","-3 32592..33221   630\n","-1 39914..40525   612\n","+1 53977..54588   612\n","+3 54588..55193   606\n","-3 33234..33809   576\n","-3 22002..22559   558\n","-3 23859..24413   555\n","-2  1945.. 2490   546\n","+1 73861..74370   510\n","-2  6214.. 6696   483\n","+1 16324..16806   483\n","-1 22556..23032   477\n","+3 32808..33260   453\n","+2 53324..53776   453\n","+1 29056..29502   447\n","-2 36286..36729   444\n","+3 51396..51833   438\n","+2 55196..55627   432\n","-1  2468.. 2896   429\n","-2 31798..32220   423\n","+1 52891..53307   417\n","-2 30595..30996   402\n","-2  5809.. 6204   396\n","+1     1..  393   393\n","-2 35740..36129   390\n","-3 34542..34925   384\n","+2 56474..56857   384\n","-1 17888..18268   381\n","-3 23004..23381   378\n","-3 52968..53336   369\n","-1 32207..32572   366\n","-1 23396..23752   357\n","+2 36419..36775   357\n","-2 18268..18609   342\n","+2 44594..44935   342\n","+1 43714..44049   336\n","+2  6557.. 6886   330\n","-2 34927..35253   327\n","-3 34131..34451   321\n","+1 30778..31095   318\n","+2 24974..25288   315\n","+1  1246.. 1557   312\n","-3 35253..35561   309\n","-2 39160..39468   309\n","-2 68932..69234   303"]},{"cell_type":"markdown","metadata":{"id":"kuXJucuXGDeE"},"source":["# findORFs output using lab5test.fa at mG=0 cutoff\n","\n","(Note: I have included the sequence of the gene candidate as an additional column)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"VMhrUuCBGDeE"},"outputs":[],"source":["test\n","+1     1..    9     9 ATGATGTAA\n","+3     1..    9     9 ATGATGTAA\n","-1     1..    9     9 TTACATCAT\n","-2     1..    9     9 TTACATCAT\n","-3     1..    9     9 TTACATCAT\n","+1     4..    9     6 ATGTAA\n","+2     1..    4     4 ATGA\n","test2\n","+1     1..   10    10 AATGATGTAA\n","+2     1..   10    10 AATGATGTAA\n","-1     1..   10    10 TTACATCATT\n","-2     1..   10    10 TTACATCATT\n","-3     1..   10    10 TTACATCATT\n","+2     2..   10     9 ATGATGTAA\n","+2     5..   10     6 ATGTAA\n","+3     1..    5     5 AATGA\n","test3\n","+2     1..   11    11 AAATGATGTAA\n","+3     1..   11    11 AAATGATGTAA\n","-1     1..   11    11 TTACATCATTT\n","-2     1..   11    11 TTACATCATTT\n","-3     1..   11    11 TTACATCATTT\n","+3     3..   11     9 ATGATGTAA\n","+1     1..    6     6 AAATGA\n","+3     6..   11     6 ATGTAA\n","test-1\n","+1     1..    9     9 TTACATCAT\n","+2     1..    9     9 TTACATCAT\n","+3     1..    9     9 TTACATCAT\n","-1     1..    9     9 ATGATGTAA\n","-3     1..    9     9 ATGATGTAA\n","-1     1..    6     6 ATGTAA\n","-2     6..    9     4 ATGA\n","test-2\n","+1     1..   10    10 TTACATCATA\n","+2     1..   10    10 TTACATCATA\n","+3     1..   10    10 TTACATCATA\n","-1     1..   10    10 TATGATGTAA\n","-2     1..   10    10 TATGATGTAA\n","-2     1..    9     9 ATGATGTAA\n","-2     1..    6     6 ATGTAA\n","-3     6..   10     5 TATGA\n","test-3\n","+1     1..   11    11 TTACATCATAA\n","+2     1..   11    11 TTACATCATAA\n","+3     1..   11    11 TTACATCATAA\n","-2     1..   11    11 TTATGATGTAA\n","-3     1..   11    11 TTATGATGTAA\n","-3     1..    9     9 ATGATGTAA\n","-3     1..    6     6 ATGTAA\n","-1     6..   11     6 TTATGA\n","test1A\n","+3     1..   10    10 ATGATGTAAA\n","-1     1..   10    10 TTTACATCAT\n","-2     1..   10    10 TTTACATCAT\n","-3     1..   10    10 TTTACATCAT\n","+1     1..    9     9 ATGATGTAA\n","+1     4..    9     6 ATGTAA\n","+2     1..    4     4 ATGA\n","test2A\n","+1     1..   11    11 AATGATGTAAA\n","-1     1..   11    11 TTTACATCATT\n","-2     1..   11    11 TTTACATCATT\n","-3     1..   11    11 TTTACATCATT\n","+2     1..   10    10 AATGATGTAA\n","+2     2..   10     9 ATGATGTAA\n","+2     5..   10     6 ATGTAA\n","+3     1..    5     5 AATGA\n","test3A\n","+2     1..   12    12 AAATGATGTAAA\n","-1     1..   12    12 TTTACATCATTT\n","-2     1..   12    12 TTTACATCATTT\n","-3     1..   12    12 TTTACATCATTT\n","+3     1..   11    11 AAATGATGTAA\n","+3     3..   11     9 ATGATGTAA\n","+1     1..    6     6 AAATGA\n","+3     6..   11     6 ATGTAA\n","test-1A\n","+1     1..   10    10 ATTACATCAT\n","+2     1..   10    10 ATTACATCAT\n","+3     1..   10    10 ATTACATCAT\n","-3     1..   10    10 ATGATGTAAT\n","-1     2..   10     9 ATGATGTAA\n","-1     2..    7     6 ATGTAA\n","-2     7..   10     4 ATGA\n","test-2A\n","+1     1..   12    12 AATTACATCATA\n","+2     1..   12    12 AATTACATCATA\n","+3     1..   12    12 AATTACATCATA\n","-1     1..   12    12 TATGATGTAATT\n","-2     3..   12    10 TATGATGTAA\n","-2     3..   11     9 ATGATGTAA\n","-2     3..    8     6 ATGTAA\n","-3     8..   12     5 TATGA\n","test-3A\n","+1     1..   13    13 AATTACATCATAA\n","+2     1..   13    13 AATTACATCATAA\n","+3     1..   13    13 AATTACATCATAA\n","-2     1..   13    13 TTATGATGTAATT\n","-3     3..   13    11 TTATGATGTAA\n","-3     3..   11     9 ATGATGTAA\n","-3     3..    8     6 ATGTAA\n","-1     8..   13     6 TTATGA"]}],"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.12.1"},"nteract":{"version":"0.28.0"},"colab":{"provenance":[]}},"nbformat":4,"nbformat_minor":0}